packages/uploadthing/test/__test-helpers.ts

```diff
@@ -94,111 +77,233 @@ const mockPresigned = (file: {
   };
 };
 
-export const mockExternalRequests =
-  (db: MockDbInterface): FetchEsque =>
-  async (_url, init) => {
-    fetchMock(_url, init);
-    if (_url instanceof Request) return new Response("Wut?", { status: 500 });
-    const url = new URL(_url);
-
-    // Pass through localhost requests
-    if (url.host.startsWith("localhost")) {
-      return fetch(_url, init);
-    }
-
-    // Mock file downloads
-    if (url.host === "cdn.foo.com") {
-      return new Response("Lorem ipsum doler sit amet", {
-        headers: { "Content-Type": "text/plain" },
-      });
-    }
-
-    // Mock UT Api
-    if (url.host === "uploadthing.com") {
-      const body = (() => {
-        if ((init?.method ?? "GET") === "GET") return null;
-        const json = JSON.parse(init?.body as string);
-        if (!isObject(json)) throw new Error("Expected body to be an object");
-        return json;
-      })();
-
-      if (url.pathname === "/api/uploadFiles") {
-        const presigneds = (body?.files as any[]).map((file) => {
+export const msw = setupServer(
+  /**
+   * S3
+   */
+  http.post("https://bucket.s3.amazonaws.com", async ({ request }) => {
+    s3Mock(request.url, {
```

Comment:
The `s3Mock()` function seems to be accepting something _very_ similar to `input, init` of a `Request` constructor. 

I suggest you align the call signature to allow passing the `request` clone here instead:

```suggestion
    s3Mock(request.clone())
```

This will forward the entire request so you don't have to unwrap it, serialize its headers, or read its body. You can do all that in the `s3Mock()` function, if needed at all.

Reply:
I really just want a way to easily assert what requests was made and what they were made with. 

I tried spy'ing on the global fetch but that didn't seem to pick up all the calls for some reason....

Reply:
@juliusmarminge, have you tried [Life-cycle events](https://mswjs.io/docs/api/life-cycle-events)? They give you the request resolution events that you can spy on. If you don't mean to affect the network, you should use life-cycle events as those are readonly but still let you know what requests/responses happened. 

```ts
// Create a spy function that gets the intercepted request
// and serializes it for easier assertion in tests.
const requestSpy = vi.fn(async ({ request }) => ({
  method: request.method,
  headers: Object.fromEntries(request.headers.entries()),
  body: await request.formData()
}))

// Let MSW tell you whenever requests happen.
// You can also use "request:handled" to spy only
// on the requests you have request handlers for.
server.events.on('request:start', requestSpy)

// Cleaning up is important.
afterEach(() => vi.resetAllMocks)

it('some test', async () => {
  await renderYourCode()

  // Assert on the spy just like you do now. 
  expect(requestSpy).toHaveBeenCalledWith({ ... })
})
```

> I tried spy'ing on the global fetch but that didn't seem to pick up all the calls for some reason....

Not sure why this is happening. Doesn't look to be related to MSW.

Reply:
Hmm your code doesn't seem to work as we're calling the spy with `Request` and can't assert on the serialized object. Tried something like this but looks like the life-cycle event listener can't be async:

![CleanShot 2024-03-24 at 11 38 46](https://github.com/pingdotgg/uploadthing/assets/51714798/7c50c719-df81-4af0-a7fe-a69b6f9ef7a2)

Reply:
with your way we can assert with `toHaveReturnedWith` instead but that reads a bit unintuitevly imo

Reply:
worked around it a bit, not the cleanest as it requires every handler to `await callRequestSpy` but it works.

Reply:
I'm quite sure mocking the implementation will work but you are right, you'd be asserting that the function _returned_ an object, wasn't called with it.